### DISCUSSION TITLE
How to extend an action/check?

### DISCUSSION THREAD

**Discussion by IlyaSemenov** [open-circle/valibot#833] [Q&A] [answered]
As a library developer, I want to extend some existing action/check with new behavior. Let's say I want to extend `url` so that it validates that URL is safe to use in browser links (prevent `javascript:alert("pwn3d")` alike input).

Is there a way to do so with minimal boilerplate? What I mean is, I can't use something like:

```ts
const safeUrl = (message?: string) => v.pipe(v.url(message), v.check(isSafeUrl, message))
```

because `pipe()` expects a schema as the first argument. Is there a similar counterpart for actions?

**Comment by fabian-hiller** [comment, marked as answer]
This probably works as expected when adding `v.string()` to the beginning of the pipeline. Without the schema, the actions wouldn't know what type to expect because they infer it from the schema. But I understand that this is not optimal in all cases and a more minimal API without a schema would be appreciated. Feel free to create an issue so others can join and discuss.

**Reply by IlyaSemenov** [reply]
Thanks, here's the issue: https://github.com/fabian-hiller/valibot/issues/835.
